Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.m

```diff
@@ -0,0 +1,311 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+
+#import "Crashlytics/Crashlytics/Components/FIRCLSCrashedMarkerFile.h"
+#import "Crashlytics/Crashlytics/Controllers/FIRCLSManagerData.h"
+#import "Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.h"
+#import "Crashlytics/Crashlytics/Helpers/FIRCLSCallStackTree.h"
+#import "Crashlytics/Crashlytics/Helpers/FIRCLSFile.h"
+#import "Crashlytics/Crashlytics/Models/FIRCLSExecutionIdentifierModel.h"
+#import "Crashlytics/Crashlytics/Models/FIRCLSInternalReport.h"
+#import "Crashlytics/Crashlytics/Public/FirebaseCrashlytics/FIRCrashlyticsReport.h"
+
+@interface FIRCLSMetricKitManager () {
+  NSLock *_mutex;
+  FBLPromise *_metricKitDataAvailable;
+  FIRCLSExistingReportManager *_existingReportManager;
+  FIRCLSFileManager *_fileManager;
+  FIRCLSManagerData *_managerData;
+}
+
+@end
+
+@implementation FIRCLSMetricKitManager
+
+- (instancetype)initWithManagerData:(FIRCLSManagerData *)managerData
+              existingReportManager:(FIRCLSExistingReportManager *)existingReportManager
+                        fileManager:(FIRCLSFileManager *)fileManager {
+  _existingReportManager = existingReportManager;
+  _fileManager = fileManager;
+  _managerData = managerData;
+  return self;
+}
+
+/*
+ * Registers the MetricKit manager to receive MetricKit reports by adding self to the
+ * MXMetricManager subscribers. Also initializes the promise that we'll use to ensure that any
+ * MetricKit report files are included in Crashylytics fatal reports. If no crash occurred on the
+ * last run of the app, this promise is immediately resolved so that the upload of any nonfatal
+ * events can proceed.
+ */
+- (void)registerMetricKitManager {
+  [[MXMetricManager sharedManager] addSubscriber:self];
+  _metricKitDataAvailable = [FBLPromise pendingPromise];
+
+  // If there was no crash on the last run of the app, then we aren't expecting a MetricKit
+  // diagnostic report and should resolve the promise immediately.
+  NSString *crashedMarkerFileName = [NSString stringWithUTF8String:FIRCLSCrashedMarkerFileName];
+  NSString *crashedMarkerFileFullPath =
+      [[_fileManager rootPath] stringByAppendingPathComponent:crashedMarkerFileName];
+  if (![_fileManager fileExistsAtPath:crashedMarkerFileFullPath]) {
+    [_mutex lock];
```

Comment:
For all these, you can use `@synchronized (self) { ... }` blocks instead of a mutex